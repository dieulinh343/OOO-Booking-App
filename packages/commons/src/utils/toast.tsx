import React from 'react';
import { toast, Icon } from '@ahaui/react';

type RenderContent = JSX.Element | string | JSX.Element[] | string[];

const showMessage = (type: string, title: string, content: RenderContent) => {
  const renderContent = (content: RenderContent) => (Array.isArray(content)
    ? (content.map(item => (
      <div
        key={typeof item === 'string' ? item : item.key}
        className="u-marginBottomExtraSmall"
      >
        {item}
      </div>
    )))
    : <span style={{ whiteSpace: 'pre-line' }}>{content}</span>
  );

  const render = (iconName: string, title: string, content: RenderContent) => (
    <div className="u-flex u-flexGrow-1">
      <div className="u-marginRightExtraSmall">
        <Icon name={iconName} size="medium" />
      </div>
      <div className="u-flexGrow-1">
        <div className="u-fontMedium u-marginBottomExtraSmall">{title}</div>
        <div className="u-textWordBreak">
          {renderContent(content)}
        </div>
      </div>
    </div>
  );

  switch (type) {
    case 'error': {
      toast.error(() => render('alert', title, content));
      break;
    }
    case 'warning':
      toast.warn(() => render('warning', title, content));
      break;
    case 'info':
      toast.info(() => render('informationCircle', title, content));
      break;
    case 'success': {
      toast.success(() => render('checkmarkCircle', title, content));
      break;
    }
    default:
  }
};

// ['a', 'b', 'c'] -> { 0: 'a', 1: 'b', 2: 'c' }
const arrayToObject = (array: any[]) => {
  let result = {};
  array.forEach((item, index) => {
    result = {
      ...result,
      [index]: item,
    };
  });
  return result;
};

/* Flatten the error into this structure:
[
  {
    name: 'long.key',
    message: 'error message',
  },
  ...
]
*/
const flattenErrorData = (data: any, prefix = ''): string | string[] => {
  if (data === null) {
    return [];
  }

  if (Array.isArray(data)) {
    if (data.every(i => (typeof i === 'string'))) {
      // If prefix is empty, display error content only
      const result = prefix ? [`[${prefix}] ${data.join('\n')}`] : data.join('\n');
      return result;
    }
    return flattenErrorData(arrayToObject(data), prefix);
  }

  if (typeof data === 'object') {
    let result: string[] = [];
    Object.keys(data).forEach((key) => {
      const childErrorMessages = flattenErrorData(data[key], prefix ? `${prefix}.${key}` : key);

      if (Array.isArray(childErrorMessages)) {
        result = [
          ...result,
          ...childErrorMessages,
        ];
      } else {
        result = [
          ...result,
          childErrorMessages,
        ];
      }
    });
    return result;
  }

  return prefix ? [`[${prefix}] ${data}`] : [data];
};

/**
 *
 * @param {string|object} error Error
 * @returns {string|array} The error message or a list of error string
 */
const getErrorMsg = (error: any) => {
  let message: string | string[] = '';

  if (
    typeof error === 'object'
    && !Array.isArray(error)
  ) {
    if (error.data && error.data.errorData && Object.keys(error.data.errorData).length > 0) {
      message = flattenErrorData(error.data.errorData);
    }

    if (message.length === 0) {
      if (error.data && error.data.errorMessage) {
        if (error.data.errorMessage === 'Failed to fetch') {
          message = 'Cannot connect to the server. Please try again later.';
        } else {
          message = error.data.errorMessage;
        }
      } else if (error.data && error.data.message) {
        if (error.data.message === 'Failed to fetch') {
          message = 'Cannot connect to the server. Please try again later.';
        } else {
          message = error.data.message;
        }
      } else if (error.message) {
        if (error.message === 'Failed to fetch') {
          message = 'Cannot connect to the server. Please try again later.';
        } else {
          message = error.message;
        }
      } else {
        message = 'Something went wrong.';
      }
    }
  } else {
    message = error || 'Something went wrong.';
  }
  return message;
};

function error(...args: any[]): void {
  if (args.length > 0 && args[0]?.name === 'ClientSessionExpiredError') {
    return;
  }

  if (args[0]?.ignore) {
    return;
  }

  switch (args.length) {
    case 1: {
      showMessage('error', 'Oops', getErrorMsg(args[0]));
      break;
    }
    case 2: {
      showMessage('error', args[1], getErrorMsg(args[0]));
      break;
    }
    default:
      showMessage('error', 'Oops', 'Something went wrong.');
  }
}

const success = (content: RenderContent, title = 'Success') => showMessage('success', title, content);

const info = (content: RenderContent, title = 'Info') => showMessage('info', title, content);

const warning = (content: RenderContent, title = 'Warning') => showMessage('warning', title, content);

export default {
  error,
  success,
  info,
  warning,
  getErrorMsg,
  arrayToObject,
  flattenErrorData,
  showMessage,
};
