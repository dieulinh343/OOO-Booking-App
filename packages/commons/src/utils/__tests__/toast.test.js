import React from 'react';
import { shallow } from 'enzyme';
import { toast as libToast } from '@ahaui/react';
import toast from '../toast';

jest.mock('@ahaui/react', () => ({
  Icon: function Icon() { return (<div>Icon</div>); },
  toast: {
    error: jest.fn(),
    success: jest.fn(),
    info: jest.fn(),
  },
}));

describe('utils/toast', () => {
  it('should toast success correctly', () => {
    toast.success('Hello', 'From the other side');
    const lastCall = libToast.success.mock.calls.pop();
    const RenderedComponent = lastCall[0];
    const wrapper = shallow(<RenderedComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toast success with array of contents correctly', () => {
    toast.success(['From the other side', 'And this side']);
    const lastCall = libToast.success.mock.calls.pop();
    const RenderedComponent = lastCall[0];
    const wrapper = shallow(<RenderedComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toast error with 1 arg correctly', () => {
    toast.error('An error');
    const lastCall = libToast.error.mock.calls.pop();
    const RenderedComponent = lastCall[0];
    const wrapper = shallow(<RenderedComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toast error with 2 args correctly', () => {
    toast.error('Hey', 'An error');
    const lastCall = libToast.error.mock.calls.pop();
    const RenderedComponent = lastCall[0];
    const wrapper = shallow(<RenderedComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toast error with none argument correctly', () => {
    toast.error();
    const lastCall = libToast.error.mock.calls.pop();
    const RenderedComponent = lastCall[0];
    const wrapper = shallow(<RenderedComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render anything if showMessage is not provided with correct type', () => {
    toast.showMessage('info', 'info', 'info detail');
    expect(libToast.error.mock.calls).toEqual([]);
  });

  it('arrayToObject', () => {
    expect(toast.arrayToObject([
      'im a string',
      'string 2',
      { key: 123 },
      [123, 456],
    ])).toEqual({
      0: 'im a string',
      1: 'string 2',
      2: { key: 123 },
      3: [123, 456],
    });
  });

  it('should flatten error correctly', () => {
    expect(toast.flattenErrorData({
      field1: {
        field11: 'error 11',
        field12: [
          'error 120',
          'error 121',
          {
            field1221: 'error 1221',
          },
        ],
        field13: null,
      },
      field2: 'error 2',
      field16: null,
    })).toEqual(
      [
        '[field1.field11] error 11',
        '[field1.field12.0] error 120',
        '[field1.field12.1] error 121',
        '[field1.field12.2.field1221] error 1221',
        '[field2] error 2',
      ],
    );
  });

  describe('should getErrorMsg correctly', () => {
    expect(toast.getErrorMsg(undefined)).toEqual('Something went wrong.');
    expect(toast.getErrorMsg(['item 1'])).toEqual(['item 1']);
    expect(toast.getErrorMsg({
      data: {
        errorMessage: 'Sample error msg',
      },
    })).toEqual('Sample error msg');
    expect(toast.getErrorMsg({
      data: {
        message: 'Sample error msg',
      },
    })).toEqual('Sample error msg');
    expect(toast.getErrorMsg({
      message: 'Sample error msg',
    })).toEqual('Sample error msg');
    expect(toast.getErrorMsg({
      not_data: null,
    })).toEqual('Something went wrong.');
    expect(toast.getErrorMsg({
      data: {
        errorData: 'Sample error msg',
      },
    })).toEqual(['Sample error msg']);
    expect(toast.getErrorMsg({
      data: {
        errorData: {
          field1: 'Sample error msg',
        },
      },
    })).toEqual(['[field1] Sample error msg']);
    expect(toast.getErrorMsg({
      data: {
        errorData: {
          field1: {
            field11: 'Sample error msg',
          },
          field2: 'Sample error msg',
        },
      },
    })).toEqual(['[field1.field11] Sample error msg',
      '[field2] Sample error msg']);
  });
});
