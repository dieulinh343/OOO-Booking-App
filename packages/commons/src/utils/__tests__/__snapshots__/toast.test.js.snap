// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/toast should toast error with 1 arg correctly 1`] = `
<div
  className="u-flex u-flexGrow-1"
>
  <div
    className="u-marginRightExtraSmall"
  >
    <Icon
      name="alert"
      size="medium"
    />
  </div>
  <div
    className="u-flexGrow-1"
  >
    <div
      className="u-fontMedium u-marginBottomExtraSmall"
    >
      Oops
    </div>
    <div
      className="u-textWordBreak"
    >
      <span
        style={
          Object {
            "whiteSpace": "pre-line",
          }
        }
      >
        An error
      </span>
    </div>
  </div>
</div>
`;

exports[`utils/toast should toast error with 2 args correctly 1`] = `
<div
  className="u-flex u-flexGrow-1"
>
  <div
    className="u-marginRightExtraSmall"
  >
    <Icon
      name="alert"
      size="medium"
    />
  </div>
  <div
    className="u-flexGrow-1"
  >
    <div
      className="u-fontMedium u-marginBottomExtraSmall"
    >
      An error
    </div>
    <div
      className="u-textWordBreak"
    >
      <span
        style={
          Object {
            "whiteSpace": "pre-line",
          }
        }
      >
        Hey
      </span>
    </div>
  </div>
</div>
`;

exports[`utils/toast should toast error with none argument correctly 1`] = `
<div
  className="u-flex u-flexGrow-1"
>
  <div
    className="u-marginRightExtraSmall"
  >
    <Icon
      name="alert"
      size="medium"
    />
  </div>
  <div
    className="u-flexGrow-1"
  >
    <div
      className="u-fontMedium u-marginBottomExtraSmall"
    >
      Oops
    </div>
    <div
      className="u-textWordBreak"
    >
      <span
        style={
          Object {
            "whiteSpace": "pre-line",
          }
        }
      >
        Something went wrong.
      </span>
    </div>
  </div>
</div>
`;

exports[`utils/toast should toast success correctly 1`] = `
<div
  className="u-flex u-flexGrow-1"
>
  <div
    className="u-marginRightExtraSmall"
  >
    <Icon
      name="checkmarkCircle"
      size="medium"
    />
  </div>
  <div
    className="u-flexGrow-1"
  >
    <div
      className="u-fontMedium u-marginBottomExtraSmall"
    >
      From the other side
    </div>
    <div
      className="u-textWordBreak"
    >
      <span
        style={
          Object {
            "whiteSpace": "pre-line",
          }
        }
      >
        Hello
      </span>
    </div>
  </div>
</div>
`;

exports[`utils/toast should toast success with array of contents correctly 1`] = `
<div
  className="u-flex u-flexGrow-1"
>
  <div
    className="u-marginRightExtraSmall"
  >
    <Icon
      name="checkmarkCircle"
      size="medium"
    />
  </div>
  <div
    className="u-flexGrow-1"
  >
    <div
      className="u-fontMedium u-marginBottomExtraSmall"
    >
      Success
    </div>
    <div
      className="u-textWordBreak"
    >
      <div
        className="u-marginBottomExtraSmall"
        key="From the other side"
      >
        From the other side
      </div>
      <div
        className="u-marginBottomExtraSmall"
        key="And this side"
      >
        And this side
      </div>
    </div>
  </div>
</div>
`;
